lib/experimental/lexopt.nim

```diff
@@ -0,0 +1,251 @@
+    cmdEnd ## End of command line
+    cmdLong ## A long option (ie. `--opt`)
+    cmdShort ## A short option (ie. `-o`)
+    cmdValue ## A value that is not an option
+
+  CmdLexer* = object
+    cmdline: seq[string] ## The command line to parse
+    index: int ## The index of `cmdline` to be processed
+    valueIdx: int ## \
+      ## For long options, the index of `ValueDelims`.
+      ##
+      ## For short options, the index of the next flag to be processed.
+      ##
+      ## Otherwise, it is unused.
+
+  LexError* = object of CatchableError
+    ## Errors occurred while lexing
+
+  UnexpectedValueError* = object of LexError
+    ## A value was provided but was not consumed
+    opt*: string ## The option that a value was provided to
+    value*: string ## The provided value
+
+proc initCmdLexer*(args: sink seq[string]): CmdLexer =
+  ## Creates a new `CmdLexer`. `args` should be the command line parameters as
+  ## returned by `os.commandLineArgs()`.
+  CmdLexer(cmdline: args)
+
+proc initCmdLexer*(args: openArray[string]): CmdLexer =
+  ## Creates a new `CmdLexer`. `args` should be the command line parameters as
+  ## returned by `os.commandLineArgs()`.
+  CmdLexer(cmdline: @args)
+
+template current(l: CmdLexer): string =
+  l.cmdline[l.index]
+
+proc isLongOpt*(s: string): bool = s.startsWith("--")
+  ## Returns whether `s` can be parsed as a long option.
+proc isShortOpt*(s: string): bool = s.startsWith('-')
+  ## Returns whether `s` can be parsed as a short option.
+proc isDash(s: string): bool = s == "-"
+proc isDashDash(s: string): bool = s == "--"
+
+proc prefix*(kind: CmdlineKind): string =
+  ## Return the prefix string for a given command line parameter kind
+  case kind
+  of cmdLong:
+    "--"
+  of cmdShort:
+    "-"
+  else:
+    ""
+
+func newUnexpectedValueError(opt, value: string): ref UnexpectedValueError =
+  (ref UnexpectedValueError)(
+    msg: "unexpected value '" & value & "' for '" & opt & "'",
+    opt: opt,
+    value: value
+  )
+
+proc next*(l: var CmdLexer): (CmdlineKind, string) =
+  ## Consume the next option.
+  ##
+  ## The following section describes how various syntaxes are handled:
+  ##
+  ## 1. `--` and `-` are considered values.
+  ## 2. `--foo:bar` and `--foo=bar` are treated as long options with key `foo`
+  ##    and `bar` is considered the value of `foo`.
+  ##
+  ##    If this value is not consumed before the next call to `next()`, an error
+  ##    will be raised. See below for more details.
+  ## 3. `--foo` is treated as a long option with key `foo`.
+  ## 4. `-abco` is treated as into `a`, `b`, `c`, `o` short options, returned
+  ##    one per `next()` call.
+  ## 5. `-abco:foo` and `-abco=foo` are treated as `a`, `b`, `c`, `o` short
+  ##    options, with `foo` being the value of option `o`.
+  ##
+  ##    If this value is not consumed before the next call to `next()`, an error
+  ##    will be raised. See below for more details
+  ## 6. Everything else is considered values.
```

Comment:
For completeness' sake, I think single short options (i.e., `-a`) should also be covered.